src/lang.rs

```diff
@@ -907,13 +907,14 @@ impl Context for CallStack {
     }
 
     fn get_mut(&mut self, name: String) -> EvalResult {
-        let (obj, env) = self.find(name.clone())?;
+        let (obj, _env) = self.find(name.clone())?;
 
-        if self.env() == env {
-            return Ok(obj);
-        }
+        let objc = match obj {
+            // when accessed mutably, promises are masked by materialized value
+            Obj::Promise(Some(x), ..) => *x.clone(),
+            _ => obj.clone(),
+        };
 
-        let objc = obj.clone();
         self.env().insert(name, objc.view_mut());
         Ok(objc)
```

Comment:
This should still work, but I think it is better to do the env comparison than the insertion into the environment which requires a hash. The env comparison should also often evaluate to true so it might be worth to have the optimisation.